resources/overlay_plugin_api.ts

```diff
@@ -0,0 +1,221 @@
+            delete responsePromises[msg.rseq];
+          } else {
+            processEvent(msg);
+          }
+        } catch (e) {
+          console.error('Invalid message received: ', _msg);
+          return;
+        }
+      });
+
+      ws.addEventListener('close', () => {
+        queue = [];
+
+        console.log('Trying to reconnect...');
+        // Don't spam the server with retries.
+        setTimeout(() => {
+          connectWs();
+        }, 300);
+      });
+    };
+
+    connectWs();
+  } else {
+    const waitForApi = function() {
+      if (!window.OverlayPluginApi || !window.OverlayPluginApi.ready) {
+        setTimeout(waitForApi, 300);
+        return;
+      }
+
+      const q = queue;
+      queue = [];
+
+      window.__OverlayCallback = processEvent;
+
+      sendMessage({
+        call: 'subscribe',
+        events: Object.keys(subscribers),
+      });
+
+      for (const item of q) {
+        if (Array.isArray(item))
+          sendMessage(item[0], item[1]);
+      }
+    };
+
+    waitForApi();
+  }
+
+  const processEvent = <T extends EventType>(msg: Parameters<EventMap[T]>[0]): void => {
+    const subs = subscribers[msg.type];
+    subs?.forEach((sub) => sub(msg as unknown));
+  };
+
+  window.dispatchOverlayEvent = processEvent;
+}
+export const addOverlayListener = <T extends EventType>(event: T, cb: EventMap[T]): void => {
+  if (kOverrides.addOverlayListenerOverride)
+    return kOverrides.addOverlayListenerOverride(event, cb);
+
+  if (!subscribers[event]) {
+    subscribers[event] = [];
+
+    if (!queue) {
+      sendMessage({
+        call: 'subscribe',
+        events: [event],
+      });
+    }
+  }
+
+  subscribers[event]?.push(cb as VoidFunc<unknown>);
+};
+
+export const removeOverlayListener = <T extends EventType>(event: T, cb: EventMap[T]): void => {
+  if (kOverrides.removeOverlayListenerOverride)
+    return kOverrides.removeOverlayListenerOverride(event, cb);
+
+  if (subscribers[event]) {
+    const list = subscribers[event];
+    const pos = list?.indexOf(cb as VoidFunc<unknown>);
+
+    if (pos && pos > -1) list?.splice(pos, 1);
+  }
+};
+
+export const callOverlayHandler: IOverlayHandler = (
+    _msg: { [s: string]: unknown },
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+): Promise<any> => {
+  if (kOverrides.callOverlayHandlerOverride)
+    return kOverrides.callOverlayHandlerOverride(_msg as Parameters<IOverlayHandler>[0]);
+
+  const msg = {
+    ..._msg,
+    rseq: 0,
+  };
+  let p: Promise<unknown>;
+
+  if (ws) {
+    msg.rseq = rseqCounter++;
+    p = new Promise((resolve) => {
+      responsePromises[msg.rseq] = resolve;
+    });
+
+    sendMessage(msg);
+  } else {
+    p = new Promise((resolve) => {
+      sendMessage(msg, (data) => {
+        resolve(data === null ? null : JSON.parse(data));
+      });
+    });
+  }
+
+  return p;
+};
+
+interface Overrides {
+  addOverlayListenerOverride?: <T extends EventType>(event: T, cb: EventMap[T]) => void;
```

Comment:
Is it possible to shorten this with `addOverlayListenerOverride: typeof addOverlayListener` sort of thing, so that the written out type is not duplicated?